### PR TITLE
Continuous Fuzzing via Fuzzit

### PR DESCRIPTION
**Pull request by yevgenypats**
This PR adds continuous fuzzing to OPA via [fuzzit](https://fuzzit.dev)

More details explained in this issue: https://github.com/open-policy-agent/opa/issues/1652

cc @tsandall

### CODE REVIEW

**Comment by yevgenypats** [comment]
@tsandall this is ready for your review.

File: .travis.yml [hunk 1]
```diff
@@ -52,3 +55,15 @@ deploy:
...
       condition: $TRAVIS_BRANCH =~ ^release-.*$
       tags: true
+
+  # This deployment is mutually exclusive with the other deployments. This
+  # deployment builds and updates the fuzzers on fuzzit whenever changes are
+  # committed to master. The edge tag is updated to track this version.
+  # The latest tag is not updated.
+  - provider: script
+    skip_cleanup: true  # do not delete artifacts because linux executable is needed
+    script: FUZZIT_TYPE=fuzzing make fuzzit
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply .travis.yml]: done

File: Makefile [hunk 2]
```diff
@@ -81,6 +81,16 @@ install: generate
...
 test: opa-wasm-test go-test wasm-test
 
+.PHONE: fuzzit
+fuzzit:
+	$(GO) get -u github.com/dvyukov/go-fuzz/go-fuzz github.com/dvyukov/go-fuzz/go-fuzz-build
+	go-fuzz-build -libfuzzer -o ast-fuzzer.a ./ast/
+	clang -fsanitize=fuzzer ast-fuzzer.a -o ast-fuzzer
+	wget -q -O fuzzit https://github.com/fuzzitdev/fuzzit/releases/download/v2.4.29/fuzzit_Linux_x86_64
+	chmod a+x fuzzit
+	./fuzzit create job --type "${FUZZIT_TYPE}" opa/ast ast-fuzzer
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply Makefile]: done

**Comment by yevgenypats** [comment]
@patrick-east I this is ready. can you login to https://app.fuzzit.dev so I can add you to opa account and you can takeover.

**Comment by patrick-east** [comment]
> can you login to https://app.fuzzit.dev so I can add you to opa account and you can takeover.

Done. We should make sure @tsandall has access too (if it isn't already set up)

**Comment by yevgenypats** [comment]
@patrick-east  you should have access now https://app.fuzzit.dev/orgs/opa/dashboard and you are admin in that account so you can add people as necessary. I'm not sure @tsandall is registered yet.

Do you think it's ready to be merged?:)

File: .travis.yml [hunk 3]
```diff
@@ -52,3 +55,15 @@ deploy:
       repo: open-policy-agent/opa
       condition: $TRAVIS_BRANCH =~ ^release-.*$
       tags: true
+
+  # This deployment is mutually exclusive with the other deployments. This
+  # deployment builds and updates the fuzzers on fuzzit whenever changes are
+  # committed to master. The edge tag is updated to track this version.
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply .travis.yml]: done

File: .travis.yml [hunk 4]
```diff
@@ -1,6 +1,7 @@
 sudo: required
 language: go
 go_import_path: github.com/open-policy-agent/opa
+dist: bionic
```
[the remark this answers is not in the document]
Comment (yevgenypats) [reply .travis.yml]: done

**Comment by tsandall** [comment]
@yevgenypats I forgot to ask. Where is the corpus coming from?

**Comment by yevgenypats** [comment]
@tsandall the corpus is generated while the fuzzers are running on fuzzit. so the corpus is download from fuzzit servers, you can download it from the UI and via the CLI.

One important note: for regression to work on PR the corpus should be public as secrets are not [accessible](https://docs.travis-ci.com/user/pull-requests#pull-requests-and-security-restrictions) in the CI from forked PR for obvious reasons. This is usually not a problem for most projects as the corpus also doesn't contain crashes that are not fixed. If this is a problem due to some highly sensitive project then I advice on activating only the fuzzing without the regression testing.

You should have access now to https://app.fuzzit.dev/orgs/opa/dashboard

**Comment by yevgenypats** [comment]
Hooray! feel free to RT https://twitter.com/fuzzitdev/status/1167127542282764288

**Pull request merged**